src/sources/socket/mod.rs

```diff
@@ -382,7 +385,7 @@ mod test {
             MaybeTlsSettings::Raw(()),
         );
         rt.spawn(
-            futures01::stream::iter_ok::<_, ()>(std::iter::repeat(()))
+            futures01::stream::iter_ok::<_, ()>(std::iter::repeat(()).take(1_000))
```

Comment:
I added `.take(1_000)` because if left unlimited stream test will fail.
Will fail because once `TcpSource` receive shutdown signal and close socket with `Shutdown::Write`, `poll_complete` on `TcpSink` side will not be ready because `mpsc` in `TcpSource` will full, as result `TcpSink` will not shutdown and `TcpSource` will live while `TcpSink` live. Looks like some deadlock. Maybe this problem go away with moving more code to new futures? Not sure.
https://github.com/timberio/vector/blob/fb97d37dc30a700bd74eda9bd37f42faa2cd654e/src/sinks/util/tcp.rs#L224-L233

Reply:
That's exactly what's happening, is what I wanted to say, but after some digging, even in that case it should have been forcibly shutdown. Something else is going on, try increasing the thread count.

Reply:
More context, `TcpSource` should live so long all it's received events haven't been processed, this is graceful shutdown, but for this case, and others, on top of graceful we have forceful which should have shutdown it.

Reply:
Maybe it's an issue of using dedicated threads for blocking operations, I don't think they can be forcibly stopped.

Reply:
I would much rather we fix the root cause here than add the `take`. We need to make sure shutdown still works properly.